CMakeLists.txt

```diff
@@ -77,100 +74,132 @@ set(USE_EMBEDDED_REVISION ON CACHE BOOL "Embed revision numbers")
 
 # Enable cross compilation for specific architectures: For eg. -march=knl for knl architecture
 set(USE_ARCH native CACHE STRING "Use specific architecture for cross compilation (Default : native)")
-set(NUM_TEST_THREADS "" CACHE STRING "Number of threads to use for running the tests.")
 
-###### Configure (users don't need to go beyond here) ######
+set(NUM_TEST_THREADS "" CACHE STRING "Maximum number of threads to use when running tests (default: number of physical cores)")
 
-enable_testing()
+if(NOT NUM_TEST_THREADS)
+  cmake_host_system_information(RESULT NUM_TEST_THREADS QUERY NUMBER_OF_PHYSICAL_CORES)
+endif()
+if(NUM_TEST_THREADS LESS_EQUAL 0)
+  set(NUM_TEST_THREADS 1)
+endif()
 
-###### Configure compiler ######
+if(ENABLE_HETERO_GALOIS)
+  set(USE_ARCH none)
+endif()
 
-#Always include debug info and aggressive warnings
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g -Wall")
-set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -Wall")
+###### Configure (users don't need to go beyond here) ######
 
-# Some apps break with -O3 on recent clang versions.
-# Tracking that down is a work in progress, so for the current release,
-# only use -O2 with clang.
-if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" AND "${CMAKE_BUILD_TYPE}" STREQUAL "Release")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O2")
-endif()
+# Enable testing and ctest integration
+include(CTest)
+
+###### Configure compiler ######
 
 # generate compile_commands.json
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
-# ICC
-if(CMAKE_CXX_COMPILER_ID MATCHES "Intel")
-  if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 19.0.1)
-    message(FATAL_ERROR "icpc must be 19.0.1 or higher. Found ${ICC_VERSION}.")
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF) #...without compiler extensions like gnu++11
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+
+# Always include debug info and aggressive warnings
+add_compile_options("$<$<COMPILE_LANGUAGE:CXX>:-g;-Wall;-Wextra;-Werror>")
+
+# GCC
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS 7)
+    message(FATAL_ERROR "gcc must be version 7 or higher. Found ${CMAKE_CXX_COMPILER_VERSION}.")
   endif()
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -wd68 -wd981 -wd383 -wd869 -wd2196 -wd279 -wd2504 -wd2943 -wd32013 -wd3373")
+
+  if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 7 AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 8)
```

Comment:
I'm a little torn about warning suppression in source code, especially since there are multiple compiler/versions. Ideally we should make our headers as clean as possible, IMHO, and leave inline suppression as a last resort.

Reply:
I've always viewed header cleanliness for stuff like this as something that has to be managed through sane use of macros. My main goal in this case is to make it so that an end user can include whatever they want from our library and not get errors/warnings even if they choose to compile their own code with `-Wextra -Werror`. My rule of thumb is that any compiler-specific configurations needed by a package belong in some well-isolated headers for that package. Build flags aren't picked up automatically by people who use a library but compiler-specific config headers are.

Reply:
I'll tackle this as a part of https://github.com/IntelligentSoftwareSystems/Galois/issues/45.